Add force_web_user_password_reset management command

dannyroberts opened the pull request:
## Product Description
There is no user-facing entrypoint being added here, but it allows a system administrator to run a management command to log out and force reset the passwords of a given set of Web Users.

## Technical Summary
https://dimagi-dev.atlassian.net/browse/SAAS-13598

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

dannyroberts committed:
Add force_web_user_password_reset management command

dannyroberts committed:
Make force_web_user_password_reset more verbose

dannyroberts committed:
Fix confusion between couch user and django user

dannyroberts committed:
Add --skip-if-reset-since option to force_web_user_password_reset

mjriley commented on corehq/apps/hqadmin/management/commands/force_web_user_password_reset.py:
```diff
@@ -0,0 +1,90 @@
+                web_users_to_reset.append(web_user)
+
+        if errors:
+            raise CommandError(f'The following errors were found in your input file:\n{errors}')
```
Nit: including errors this way may not be the most readable (assuming the output gets put on the command line). It might make sense to join the errors array with newlines

mjriley commented on corehq/apps/hqadmin/management/commands/force_web_user_password_reset.py:
```diff
@@ -0,0 +1,90 @@
+    user.set_password(uuid.uuid4().hex)
+    user.save()
+    url = f"{get_url_base()}{reverse('password_reset_email')}"
+    send_html_email_async.delay(
+        'Reset Password on CommCare HQ',
+        web_user.get_email(),
+        (f'Your system administrator has forced a password reset. '
+         f'Before you will be able to continue to use CommCare, '
+         f'you must reset your password by navigating to {url} and following the instructions.'),
+        email_from=settings.DEFAULT_FROM_EMAIL
+    )
```
Do we need to think about the scenario (assuming it exists) where the user is not linked to a valid email address? `get_email()` returns the email, if set, or the username. Neither of those are guaranteed to be a valid email. Presumably, a valid email was established initially, to verify the account, but after account creation, that can change. If we fail at delivering a password reset notification, are we concerned that we'll be locking out these users with no notification, or can we just expect them to contact their managers?

dannyroberts replied:
Users are expected to maintain a working email (remember these are people doing this professionally for their job), so if their email is not working then locking them out forever is actually a very sensical thing to do in this case where the program managers are concerned the credentials may have been compromised. And like you said their recourse would be to contact their managers who would sort it out with support.

mjriley commented on corehq/apps/hqadmin/management/commands/force_web_user_password_reset.py:
```diff
@@ -0,0 +1,90 @@
+            print("No web users to reset passwords for. No action taken.")
+            return
+
+        self.stdout.write("The following users will be logged out and have their passwords force reset.")
```
Does `set_password()` log users out? I'm not seeing any code that logs users out of their accounts

dannyroberts replied:
It does yeah. I've tested this on staging and it does so very reliably!

mjriley commented on corehq/apps/hqadmin/management/commands/force_web_user_password_reset.py:
```diff
@@ -0,0 +1,90 @@
+
+def force_password_reset(web_user):
+    user = web_user.get_django_user()
+    user.set_password(uuid.uuid4().hex)
```
Will this update the `last_password_set` field on the couch user? That appears to be a couch-specific field, but we're only updating a django model here.

dannyroberts replied:
I'm intentionally not updating that field, because I'm treating this operation not as us "resetting their password" but as us "unsetting" their password to force them to reset it. Until they actually reset it, I want that field to read the date of the last time their password was set (to something the client/user controls).

mjriley reviewed (approved):
Nice clean code. I think this is good enough to approve, although I'm curious about whether we update password timestamps with this script.

dannyroberts committed:
Format errors in force_web_user_password_reset more nicely

dannyroberts committed:
Explicitly use urandom rather than uuid4 for cryptographic randomness

uuid4 behind the scenes does use urandom in the python implementation we are using
but this is undocumented behavior, and UUIDs are not meant for cryptographic use
and we do not want to rely on this undocumented behavior

dannyroberts commented on corehq/apps/hqadmin/management/commands/force_web_user_password_reset.py:
```diff
@@ -0,0 +1,92 @@
+import uuid
```
```suggestion
```

dannyroberts committed:
Remove unused import

dannyroberts commented on corehq/apps/hqadmin/management/commands/force_web_user_password_reset.py:
```diff
@@ -0,0 +1,91 @@
+            raise CommandError(f'The following errors were found in your input file:\n{errors_string}')
+
+        if web_users_to_skip_due_to_recent_reset:
+            self.stdout.write("The following users will be skipped "
+                              "because they have reset their passwords since {skip_if_reset_since}")
```
```suggestion
            self.stdout.write(f"The following users will be skipped "
                              f"because they have reset their passwords since {skip_if_reset_since}")
```

dannyroberts committed:
Use f-string where originally intended

mjriley commented on corehq/apps/hqadmin/views/users.py:
```diff
@@ -398,7 +399,7 @@ def form_valid(self, form):
         reset_password = form.cleaned_data['reset_password']
         if reset_password:
-            self.user.set_password(uuid.uuid4().hex)
+            unset_password(self.user)
```
The language here seems a little confusing. 'Reset password', to me, means that the password is being changed with the intent of delivering that password to the user. It's likely not a huge deal, but I feel like the most appropriate fix is probably changing the language on this form to 'clear password' or something?

dannyroberts replied:
I think I'm going to leave that for whenever the next time we're working on this particular form is, since this was peripheral to the main change I was trying to make
